lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,417 @@
+      }
+
+      cachedPaths = new Map()
+      this.#data.set(req.origin, cachedPaths)
+    }
+
+    let values = cachedPaths.get(`${req.path}:${req.method}`)
+    if (!values && makeIfDoesntExist) {
+      values = []
+      cachedPaths.set(`${req.path}:${req.method}`, values)
+    }
+
+    return values
+  }
+
+  /**
+   * Given a list of values of a certain request, this decides the best value
+   *  to respond with.
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @param {MemoryStoreValue[]} values
+   * @returns {MemoryStoreValue | undefined}
+   */
+  #findValue (req, values) {
+    /**
+     * @type {MemoryStoreValue}
+     */
+    let value
+    const now = Date.now()
+    for (let i = values.length - 1; i >= 0; i--) {
+      const current = values[i]
+      const currentCacheValue = current.opts
+      if (now >= currentCacheValue.deleteAt) {
+        // We've reached expired values, let's delete them
+        this.#entryCount -= values.length - i
+        values.length = i
+        break
+      }
+
+      let matches = true
+
+      if (currentCacheValue.vary) {
+        if (!req.headers) {
+          matches = false
+          break
+        }
+
+        for (const key in currentCacheValue.vary) {
+          if (currentCacheValue.vary[key] !== req.headers[key]) {
+            matches = false
+            break
+          }
+        }
+      }
+
+      if (matches) {
+        value = current
+        break
+      }
+    }
+
+    return value
+  }
+}
+
+class MemoryStoreReadableStream extends Readable {
+  /**
+   * @type {MemoryStoreValue}
+   */
+  #value
+  /**
+   * @type {Buffer[]}
+   */
+  #chunksToSend = []
+
+  /**
+   * @param {MemoryStoreValue} value
+   */
+  constructor (value) {
+    super()
+
+    if (value.readLock) {
+      throw new Error('can\'t read a locked value')
+    }
+
+    this.#value = value
+    this.#chunksToSend = [...this.#value.body, null]
+
+    this.#value.readers++
+    this.#value.writeLock = true
+
+    this.on('close', () => {
+      this.#value.readers--
+
+      if (this.#value.readers === 0) {
+        this.#value.writeLock = false
+      }
+    })
+  }
+
+  get value () {
+    return this.#value.opts
+  }
+
+  /**
+   * @param {number} size
+   */
+  _read (size) {
+    if (this.#chunksToSend.length === 0) {
+      throw new Error('no chunks left to read, stream should have closed')
+    }
+
+    if (size > this.#chunksToSend.length) {
+      size = this.#chunksToSend.length
+    }
+
+    for (let i = 0; i < size; i++) {
+      this.push(this.#chunksToSend.shift())
+    }
+  }
+}
+
+class MemoryStoreWritableStream extends Writable {
+  /**
+   * @type {MemoryStoreValue}
+   */
+  #value
+  #currentSize = 0
+  #maxEntrySize = 0
+  /**
+   * @type {Buffer[]}
+   */
+  #body = []
+
+  /**
+   * @param {MemoryCacheStore} value
+   * @param {number} maxEntrySize
+   */
+  constructor (value, maxEntrySize) {
+    super()
+    this.#value = value
```

Comment:
Type 'MemoryCacheStore' is missing the following properties from type 'MemoryStoreValue': readers, readLock, writeLock, opts, bodyts(2739)